File: safe-online-sdk/src/main/java/net/link/safeonline/sdk/ws/notification/consumer/NotificationConsumerClientImpl.java
/*
 * SafeOnline project.
 *
 * Copyright 2006-2007 Lin.k N.V. All rights reserved.
 * Lin.k N.V. proprietary/confidential. Use is subject to license terms.
 */

package net.link.safeonline.sdk.ws.notification.consumer;

import com.sun.xml.ws.client.ClientTransportException;
import java.security.PrivateKey;
import java.security.cert.X509Certificate;
import javax.xml.ws.BindingProvider;
import net.lin_k.safe_online.notification.consumer.NotificationConsumerPort;
import net.lin_k.safe_online.notification.consumer.NotificationConsumerService;
import net.lin_k.safe_online.notification.consumer.NotificationMessageHolderType;
import net.lin_k.safe_online.notification.consumer.Notify;
import net.lin_k.safe_online.notification.consumer.NotificationMessageHolderType.Message;
import net.link.safeonline.notification.consumer.ws.NotificationConsumerServiceFactory;
import net.link.safeonline.sdk.logging.exception.WSClientTransportException;
import net.link.safeonline.sdk.ws.AbstractWSClient;
import net.link.util.ws.pkix.wssecurity.WSSecurityClientHandler;
import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.oasis_open.docs.wsn.b_2.TopicExpressionType;


/**
 * Implementation of the WS-Notification consumer interface. This class is using JAX-WS, secured via WS-Security and server-side SSL.
 *
 * @author wvdhaute
 */
public class NotificationConsumerClientImpl extends AbstractWSClient implements NotificationConsumerClient {

    private static final Log LOG = LogFactory.getLog( NotificationConsumerClientImpl.class );

    private static final String TOPIC_DIALECT_SIMPLE = "http://docs.oasis-open.org/wsn/t-1/TopicExpression/Simple";

    private final NotificationConsumerPort port;

    private final String location;


    /**
     * Main constructor.
     *
     * @param location the location (host:port) of the attribute web service.
     * @param clientCertificate the X509 certificate to use for WS-Security signature.
     * @param clientPrivateKey the private key corresponding with the client certificate.
     * @param serverCertificate the X509 certificate of the server
     * @param maxOffset the maximum offset of the WS-Security timestamp received. If <code>null</code> default offset configured in
     *            {@link WSSecurityClientHandler} will be used.
     * @param sslCertificate If not <code>null</code> will verify the server SSL {@link X509Certificate}.
     */
    public NotificationConsumerClientImpl(String location, X509Certificate clientCertificate, PrivateKey clientPrivateKey,
                                          X509Certificate serverCertificate, Long maxOffset, X509Certificate sslCertificate) {

        NotificationConsumerService consumerService = NotificationConsumerServiceFactory.newInstance();
        port = consumerService.getNotificationConsumerPort();
        this.location = location;
        setEndpointAddress();

        registerMessageLoggerHandler( port );

        registerTrustManager( port, sslCertificate );

        WSSecurityClientHandler.addNewHandler( port, clientCertificate, clientPrivateKey, serverCertificate, maxOffset );
    }

    private void setEndpointAddress() {

        BindingProvider bindingProvider = (BindingProvider) port;

        bindingProvider.getRequestContext().put( BindingProvider.ENDPOINT_ADDRESS_PROPERTY, location );
    }

    public void sendNotification(String topic, String destination, String subject, String content)
            throws WSClientTransportException {

        LOG.debug( "send notification to " + location + " for topic: " + topic + ", destination:" + destination + ", subject:" + subject
                + ", content:" + content );

        TopicExpressionType topicExpression = new TopicExpressionType();
        topicExpression.setDialect( TOPIC_DIALECT_SIMPLE );
        topicExpression.getContent().add( topic );

        Notify notifications = new Notify();
        NotificationMessageHolderType notification = new NotificationMessageHolderType();
        notification.setTopic( topicExpression );
        Message notificationMessage = new Message();
        notificationMessage.setDestination( destination );
        notificationMessage.setSubject( subject );
        notificationMessage.setContent( content );
        notification.setMessage( notificationMessage );
        notifications.getNotificationMessage().add( notification );

        try {
            port.notify( notifications );
        } catch (ClientTransportException e) {
            LOG.debug( "Failed to send notification" );
            throw new WSClientTransportException( location, e );
        }
    }
}

File: safe-online-sdk-dotnet/src/main/linkid-sdk-dotnet/linkid-sdk-dotnet/src/main/cs/ws/notification/NotificationConsumerClientImpl.cs
/*
 * SafeOnline project.
 * 
 * Copyright 2006-2008 	Lin.k N.V. All rights reserved.
 * Lin.k N.V. proprietary/confidential. Use is subject to license terms.
 */

using NotificationConsumerWSNameSpace;
using System;
using System.Net;
using System.Net.Security;
using System.Security.Cryptography.X509Certificates;
using System.ServiceModel;
using System.ServiceModel.Channels;
using System.ServiceModel.Description;
using System.ServiceModel.Dispatcher;
using System.ServiceModel.Security;
using System.ServiceModel.Security.Tokens;
using System.Xml;

namespace safe_online_sdk_dotnet
{
	/// <summary>
	/// Client implementation of the linkID WS-Notification Consumer Web Service.
	/// </summary>
	public class NotificationConsumerClientImpl : NotificationConsumerClient
	{
		private NotificationConsumerPortClient client;

        public NotificationConsumerClientImpl(string location, X509Certificate2 appCertificate, X509Certificate2 linkidCertificate)
		{
			string address = "https://" + location + "/linkid-ws/consumer";
			EndpointAddress remoteAddress = new EndpointAddress(address);
					
			this.client = new NotificationConsumerPortClient(new LinkIDBinding(linkidCertificate), remoteAddress);
			
			this.client.ClientCredentials.ClientCertificate.Certificate = appCertificate;
			this.client.ClientCredentials.ServiceCertificate.DefaultCertificate = linkidCertificate;
			// To override the validation for our self-signed test certificates
			this.client.ClientCredentials.ServiceCertificate.Authentication.CertificateValidationMode = X509CertificateValidationMode.None;
			
			this.client.Endpoint.Contract.ProtectionLevel = ProtectionLevel.Sign;
			this.client.Endpoint.Behaviors.Add(new LoggingBehavior());
		}
		
		public void sendNotification(string topic, string destination, string subject, string content) {
			Console.WriteLine("send notification for topic " + topic + " to destination " + destination +
			                  " for subject " + subject);
			TopicExpressionType topicExpression = new TopicExpressionType();
			topicExpression.Dialect = NotificationServiceConstants.TOPIC_DIALECT_SIMPLE;
			XmlDocument d = new XmlDocument();
			topicExpression.Any = new XmlNode[1];
			topicExpression.Any[0] = d.CreateTextNode(topic);
			
			Notify notifications = new Notify();
			NotificationMessageHolderType notification = new NotificationMessageHolderType();
			notification.Topic = topicExpression;
			
			NotificationMessageHolderTypeMessage message = new NotificationMessageHolderTypeMessage();
			message.Destination = destination;
			message.Subject = subject;
			message.Content = content;
			notification.Message = message;
			notifications.NotificationMessage = new NotificationMessageHolderType[]{notification};
			
			this.client.Notify(notifications);
		}
		
	}
}
